### PR TITLE
FIX-#4355: Fix rename algebraic operator to avoid copying

### LINKED ISSUE #4355: PERF: add_suffix and add_prefix shouldn't map partitions
### Describe the problem
`add_suffix` and `add_prefix` call the modin frame's `rename`, which [maps all blocks](https://github.com/modin-project/modin/blob/49f47b1ed66583d4b90d3c06faf3e1d457cf4df8/modin/core/dataframe/pandas/dataframe/dataframe.py#L1719). Instead, we can just change the names of the indices entire pandas object, and defer the index renaming in the partitions till later.

### PR DESCRIPTION
**Pull request by devin-petersohn**
Resolves #4355

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #4355 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by YarShev** [comment]
Some of the jobs failed. Look at them please.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -289,10 +289,24 @@ def dtypes(self):
...
 
     def add_suffix(self, suffix, axis=1):
-        return self.__constructor__(self._modin_frame.add_suffix(suffix, axis))
+        if axis == 1:
+            return self.__constructor__(
+                self._modin_frame.add_suffix(new_col_labels=lambda x: f"{x}suffix")
+            )
+        else:
+            return self.__constructor__(
+                self._modin_frame.add_suffix(new_row_labels=lambda x: f"{x}suffix")
```
Comment (mvashishtha) [inline modin/core/storage_formats/pandas/query_compiler.py:308]:
```suggestion
                self._modin_frame.rename(new_row_labels=lambda x: x + suffix)
```

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 2]
```diff
@@ -289,10 +289,24 @@ def dtypes(self):
 
     # Metadata modification methods
     def add_prefix(self, prefix, axis=1):
-        return self.__constructor__(self._modin_frame.add_prefix(prefix, axis))
+        if axis == 1:
+            return self.__constructor__(
+                self._modin_frame.rename(new_col_labels=lambda x: f"prefix{x}")
```
Comment (mvashishtha) [inline modin/core/storage_formats/pandas/query_compiler.py:294]:
```suggestion
                self._modin_frame.rename(new_col_labels=lambda x: prefix + x)
```

**Review by mvashishtha — COMMENTED** [review]
I think my suggested changes should fix the CI failures.

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 3]
```diff
@@ -1686,45 +1632,16 @@ def rename(
         -----
         If level is not specified, the default behavior is to replace row labels in all levels.
         """
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:1634]: probably should remove this note as we removed the `level` parameter

Comment (anmyachev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: Removed

Comment (anmyachev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: @dchigarev ready for review

**Comment by anmyachev** [comment]
The acceleration here can only be seen when there are several operations, due to the fact that the operation has become lazy.
For example, in such a synthetic case, the acceleration is 2 times (Ray, 8 cores).

```python
import numpy as np
import modin.pandas as pd
from modin.utils import execute
from time import time


df1 = pd.DataFrame(np.random.rand(10**6, 64))
df2 = pd.DataFrame(np.random.rand(10**6, 64))

for _ in range(5):
    start = time()
    df1_suffix = df1.add_suffix("suffix")
    res = df2.add_suffix("suffix")
    execute(res)
    print(f"add_suffix: {time()-start}")

**Comment by anmyachev** [comment]
@dchigarev please take a look

**Review by dchigarev — APPROVED** [review]

**Comment by dchigarev** [comment]
btw, why unidist tests are always failing due to timeout? maybe we should disable the hanging tests in a separate PR? cc @YarShev

**Comment by anmyachev** [comment]
> btw, why unidist tests are always failing due to timeout? maybe we should disable the hanging tests in a separate PR? cc @YarShev

@dchigarev @YarShev we could try to use https://pypi.org/project/pytest-retry/ and https://pypi.org/project/pytest-timeout/ here. Something like this:

```python
@pytest.mark.flaky(retries=3, delay=1)
@pytest.mark.timeout(60)
def test_to_csv(...):
```

I would not like to disable tests for `to_csv` on Unidist.

**Comment by anmyachev** [comment]
@YarShev @dchigarev https://github.com/modin-project/modin/pull/6776

**Pull request merged**